routes/logger.go

```diff
@@ -0,0 +1,69 @@
+package routes
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"time"
+
+	"github.com/gocardless/draupnir/routes/chain"
+	"github.com/prometheus/common/log"
+)
+
+type key int
+
+const loggerKey key = 1
+
+func NewRequestLogger(logger log.Logger) chain.Middleware {
+	return func(next chain.Handler) chain.Handler {
+		return func(w http.ResponseWriter, r *http.Request) error {
+			// To capture the response, we replace the response writer with a response
+			// recorder.
+			recorder := httptest.NewRecorder()
+
+			scopedLogger := logger.With("http_request", r)
+
+			// Inject the logger into the request's context
+			ctx := r.Context()
+			r = r.WithContext(context.WithValue(ctx, loggerKey, &scopedLogger))
+
+			// Call the next middleware and time it
+			start := time.Now()
+			next(recorder, r)
+			duration := time.Since(start)
+
+			requestLine := fmt.Sprintf(
+				"%s %s %d %f\n",
+				r.Method,
+				r.URL.String(),
+				recorder.Code,
+				duration.Seconds(),
+			)
+
+			scopedLogger.
+				With("method", r.Method).
+				With("path", r.URL.String()).
+				With("status", recorder.Code).
+				With("duration", duration.Seconds()).
+				Info(requestLine)
+
+			// Copy the headers and body from the recorder to the response writer
+			for k, v := range recorder.HeaderMap {
+				w.Header()[k] = v
+			}
+			w.WriteHeader(recorder.Code)
+			recorder.Body.WriteTo(w)
+			return nil
+		}
+	}
+}
+
+func GetLogger(r *http.Request) (log.Logger, error) {
```

Comment:
Is there anywhere that we call this function where we would expect an error? Where it would be ridiculous for us to get an error?

Because if there isn't, then I wouldn't mind panicking if the logger isn't here, and dropping the error return value. It would clean up a lot of the calling code. I'm on the fence about it but worth considering.

Reply:
Arguing devils advocate, if it panics then systemd would just bring it back up (or it should, if it doesn't).

Think it's fine for now. I'm not sure how I feel about panics in Go yet anyway, though I feel judicious use might lead to slightly cleaner code.